parmed/openmm/parameters.py

```diff
@@ -166,6 +176,13 @@ def write(self, dest, provenance=None):
             if own_handle:
                 dest.close()
 
+    def _clean_params(self):
+        keep_types = set()
+        for name, residue in iteritems(self.residues):
+            for atom in residue.atoms:
+                keep_types.add(atom.type)
+        self.skip_types = [typ for typ in self.atom_types if typ not in keep_types]
```

Comment:
Let's not cache this.  Let's have `_clean_params` (maybe rename it to `_find_unused_types`) simply return a set (which is a lot more efficient for lookups than a list).  So replace this last line with:

```Python
return {typ for typ in self.atom_types if typ not in keep_types}
```

Set comps are supported in every version of Python ParmEd supports, so this is a fine construct to use.  This will protect us against caching a list of types that may wind up out-of-date if we add any residues to our `OpenMMParameterSet`.

Then call this function right before it's used.  You can avoid calling it in all of the `_write_omm*` methods by simply calling it in `write` and passing it as a parameter -- there is absolutely no importance put on ensuring backwards-compatibility of private methods, so go ahead and make that a required positional argument for all of the `_write_omm*` methods.